packages/grid/_modules_/grid/hooks/features/core/useGridState.ts

```diff
@@ -2,30 +2,89 @@ import * as React from 'react';
 import { GRID_STATE_CHANGE } from '../../../constants/eventsConstants';
 import { GridApiRef } from '../../../models/api/gridApiRef';
 import { GridStateChangeParams } from '../../../models/params/gridStateChangeParams';
+import { isDeepEqual } from '../../../utils/utils';
 import { GridState } from './gridState';
 import { useGridApi } from './useGridApi';
 
 export const useGridState = (
   apiRef: GridApiRef,
 ): [GridState, (stateUpdaterFn: (oldState: GridState) => GridState) => boolean, () => void] => {
   useGridApi(apiRef);
+
   const forceUpdate = React.useCallback(
     () => apiRef.current.forceUpdate(() => apiRef.current.state),
     [apiRef],
   );
+
   const setGridState = React.useCallback(
     (stateUpdaterFn: (oldState: GridState) => GridState) => {
       const newState = stateUpdaterFn(apiRef.current.state);
       const hasChanged = apiRef.current.state !== newState;
+      let shouldUpdate = true;
+      const updatedStateIds: string[] = [];
+      const controlStateMap = apiRef.current.controlStateRef.current!;
+
+      if (hasChanged && apiRef.current.controlStateRef) {
+        Object.keys(controlStateMap).forEach((stateId) => {
+          const controlState = controlStateMap[stateId];
+          const oldState = controlState.stateSelector(apiRef.current.state);
+          const newSubState = controlState.stateSelector(newState);
+          const hasSubStateChanged = !isDeepEqual(oldState, newSubState);
```

Comment:
A performance bottleneck
```suggestion
          const hasSubStateChanged = oldState === newSubState;
```

Reply:
we can't === here at this stage due to the mapStateToModel that does an object.values on the state so === will never be equal. So atm if you have no mapStateToModel then it will use the === otherwise deep equal.

Reply:
looking to refactor